### ISSUE TITLE
Refactor: Implement Strategy Pattern for T-Norms (Prerequisite for Rule Vectorization)

### ISSUE THREAD

**Issue by dcruzf** [dcruzf/highFIS#54]
Currently, the T-norm operations (Product, Minimum, Geometric Mean, Dombi) are implemented as pure Python functions in highfis/t_norms.py. This functional approach introduces two significant architectural constraints:
 1. **Serialization and State Issues:** For stateful T-norms (like Dombi, which requires a lambda_ hyperparameter), the models currently rely on functools.partial to bind the parameters during initialization in highfis/models.py. Using partial inside nn.Module breaks PyTorch's serialization (torch.save / TorchScript) and prevents these hyperparameters from eventually becoming learnable nn.Parameters.
 2. **Hardcoded Dimensionality:** The current functions hardcode the reduction dimension (e.g., dim=1 in t_norm_prod).
## 🔗 Critical Prerequisite for REF-01 (Rule Vectorization)
This refactoring is a **strict prerequisite** for vectorizing the RuleLayer.
In the upcoming REF-01 vectorization, the RuleLayer will stop evaluating rules one by one. Instead, it will fetch all antecedents simultaneously, shifting the input tensor shape from (batch_size, n_inputs) to a 3D matrix of shape (batch_size, n_rules, n_inputs).
If we attempt to pass this 3D tensor to the current T-norm functions, the hardcoded dim=1 will collapse the n_rules dimension instead of the n_inputs dimension, completely destroying the rule evaluation logic. By migrating to a Strategy pattern, we can expose the reduction dim as a parameter, allowing the vectorized layer to dynamically dictate how the T-norm should behave.
## 💡 Proposed Solution
We should implement the **Strategy Pattern** by refactoring the T-norms into classes that inherit from nn.Module.
### Implementation Example
```python
import torch
from torch import Tensor, nn
from abc import ABC, abstractmethod

# 1. Base Strategy Interface
class BaseTNorm(nn.Module, ABC):
    @abstractmethod
    def forward(self, terms: Tensor, dim: int = -1) -> Tensor:
        """
        Applies the T-norm aggregation.
        The 'dim' parameter allows flexibility for 2D or 3D rule evaluations.
        """
        pass

# 2. Concrete Stateless Strategy
class ProductTNorm(BaseTNorm):
    def forward(self, terms: Tensor, dim: int = -1) -> Tensor:
        return torch.prod(terms, dim=dim)

# 3. Concrete Stateful Strategy
class DombiTNorm(BaseTNorm):
    def __init__(self, lambda_: float = 1.0, eps: float = 1e-8):
        super().__init__()
        # State is now safely tracked by the module (can be made a parameter later)
        self.lambda_ = float(lambda_)
        self.eps = eps

    def forward(self, terms: Tensor, dim: int = -1) -> Tensor:
        clamped = terms.clamp(min=self.eps, max=1.0)
        inv = (1.0 / clamped) - 1.0
        powered = torch.pow(inv, self.lambda_)
        summed = powered.sum(dim=dim)
        return 1.0 / (1.0 + torch.pow(summed, 1.0 / self.lambda_))

```
## ✅ Expected Benefits
 * **Unblocks Vectorization (REF-01):** Allows dynamic dimension reduction (dim=-1 or dim=2), making 3D tensor processing possible.
 * **Robust Serialization:** nn.Module objects serialize perfectly in PyTorch, unlike functools.partial.
 * **Future-Proofing:** Paves the way for making T-norm hyperparameters differentiable and learnable via gradient descent.
